### PR TITLE
Refactor taglist_t class

### PR DESCRIPTION
**Pull request by joto**
Removes some dodgy code including a const_cast and generally cleans up
the class.

### CODE REVIEW

File: src/osmtypes.hpp [hunk 1]
```diff
@@ -74,106 +76,162 @@ struct tag_t
...
-    tag_t *find(const std::string &key)
+    tag_t const &operator[](std::size_t idx) const noexcept
     {
-        return const_cast<tag_t *>(_find(key));
+        return m_tags[idx];
     }
 
-    int indexof(const std::string &key) const
+    /// Is there a tag with this key in the list?
+    bool contains(const std::string &key) const
```
Comment (lonvia) [inline src/osmtypes.hpp]: style: const

File: src/osmtypes.hpp [hunk 2]
```diff
@@ -74,106 +76,162 @@ struct tag_t
...
         } else {
-            tag_in_list->value = t.value;
+            it->value = std::forward<V>(value);
         }
     }
-    bool contains(const std::string &key) const { return _find(key) != 0; }
 
 private:
-    const tag_t *_find(const std::string &key) const
+    iterator find_by_key(const std::string &key)
```
Comment (lonvia) [inline src/osmtypes.hpp]: style: const

File: src/osmtypes.hpp [hunk 3]
```diff
@@ -74,106 +76,162 @@ struct tag_t
...
     {
-        for (base_t::const_iterator it = begin(); it != end(); ++it)
-            if (it->key == key)
-                return &(*it);
+        return std::find_if(m_tags.begin(), m_tags.end(),
+                            [&key](tag_t const &t) { return t.key == key; });
+    }
 
-        return 0;
+    const_iterator find_by_key(const std::string &key) const
```
Comment (lonvia) [inline src/osmtypes.hpp]: style: const

Comment (joto) [reply src/osmtypes.hpp]: all done

**Pull request merged**